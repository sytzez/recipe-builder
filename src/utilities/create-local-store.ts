import { createStore, SetStoreFunction, Store } from "solid-js/store";
import { createEffect } from "solid-js";

export function createLocalStore<T extends object>(
  name: string,
  initialState: T,
): [Store<T>, SetStoreFunction<T>] {
  const localState = localStorage.getItem(name)

  const [state, setState] = createStore<T>(
    localState ? JSON.parse(localState) : initialState
  )

  createEffect(() => localStorage.setItem(name, JSON.stringify(state)))

  return [state, setState]
}
