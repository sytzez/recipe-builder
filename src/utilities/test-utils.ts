export const waitForEvent = async (element: Element, event: string) => (
  new Promise<void>((resolve) => {
    const handler = () => {
      element.removeEventListener(event, handler)
      resolve()
    }
    element.addEventListener(event, handler)
  })
)
