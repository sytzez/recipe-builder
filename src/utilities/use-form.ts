import { createStore, SetStoreFunction, Store } from 'solid-js/store'
import { ISchema, ValidationError } from 'yup'
import { createSignal } from 'solid-js'

export const useForm = <Fields extends object>(
  initialFields: Fields,
  schema: ISchema<any>,
  submitCallback: (f: Fields) => void,
) => {
  const [fields, setFields] = createStore<Fields>(initialFields)
  const [validationError, setValidationError] =
    createSignal<ValidationError | null>(null)

  return {
    fields,
    setFields,
    setFieldUsingEvent: (fieldName: keyof Fields) => (event: Event) => {
      const inputElement = event.currentTarget as HTMLInputElement

      // @ts-ignore This should work because fieldName is a key of Fields
      setFields(fieldName, inputElement.value)
    },
    onSubmit: async (event: SubmitEvent) => {
      event.preventDefault()
      try {
        submitCallback(await schema.validate(fields))
      } catch (e) {
        setValidationError(e)
      }
    },
    validationError,
  }
}
