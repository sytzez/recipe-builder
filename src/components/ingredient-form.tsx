import { Component, createSignal, Show } from "solid-js";
// import { Ingredient } from "../types/ingredient";
import { UnitType } from "../schemata/unit-type";
import { Cost } from "../types/cost";
import { useForm } from "../utilities/use-form";
import { TextInput } from "./form/text-input";
import { ingredientSchema } from "../schemata/ingredient";
import { ValidationError } from "yup";

export interface IngredientFormProps {
    ingredient: Ingredient | null
    title: string
    submitLabel: string
    onSubmit: (Ingredient) => void
    onCancel: () => void
}

interface Fields {
    name: string
    unitType: UnitType | ''
    cost: Cost | ''
}

const emptyFields: Fields = {
    name: '',
    unitType: '',
    cost: '',
}

export const IngredientForm: Component = (props: IngredientFormProps) => {
    const initialFields = props.ingredient || emptyFields

    const { setField, onSubmit, validationError } = useForm<Fields>({ ...initialFields }, ingredientSchema, props.onSubmit, setError)

    return (
        <form
            onSubmit={onSubmit}
            class="bg-white shadow-md rounded p-8 mb-4"
        >
            <h1 class="font-bold text-2xl text-gray-800 mb-4">{props.title}</h1>
            <TextInput
                name="name"
                label="Name"
                type="text"
                placeholder="Honey"
                initialValue={initialFields.name}
                onChange={setField('name')}
                required
            />
            <TextInput
                name="unitType"
                label="Unit of measurement"
                type="text"
                placeholder="Teaspoon"
                initialValue={initialFields.unitType}
                onChange={setField('unitType')}
                required
            />
            <TextInput
                name="cost"
                label="Cost per unit of measurement"
                type="number"
                placeholder="0.05"
                initialValue={initialFields.cost}
                onChange={setField('cost')}
                required
                step=".01"
            />
            <div class="mb-4">
                <input
                    type="submit"
                    value={props.submitLabel}
                    className="bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 mr-2 rounded focus:outline-none focus:shadow-outline"
                />
                <button
                    onClick={props.onCancel}
                    className="bg-gray-100 hover:bg-gray-300 text-black font-bold py-2 px-4 rounded focus:outline-none focus:shadow-outline"
                >
                    Cancel
                </button>
            </div>
            <Show when={validationError()}>
                {(error) => <p class="text-red-600 font-bold">{error.errors.join('. ')}</p>}
            </Show>
        </form>
    )
}
