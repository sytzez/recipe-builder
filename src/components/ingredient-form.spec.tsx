import { describe, expect, it, vi } from "vitest";
import { fireEvent, render } from "solid-testing-library";
import { IngredientForm } from "./ingredient-form";
import { waitForEvent } from "../utilities/test-utils";

describe('<IngredientForm />', () => {
  const setup = () => {
    const onSubmitMock = vi.fn()

    const rendered = render(() =>
      <IngredientForm
        ingredient={null}
        title="New ingredient"
        submitLabel="Create ingredient"
        onSubmit={onSubmitMock}
      />
    )

    const submitButton = rendered.queryByText('Create ingredient')!
    const nameInput = rendered.queryByLabelText<HTMLInputElement>('Name')!
    const unitTypeInput = rendered.queryByLabelText<HTMLSelectElement>('Unit of measurement')!
    const costInput = rendered.queryByLabelText<HTMLInputElement>('Cost per unit of measurement')!

    return { ...rendered, onSubmitMock, submitButton, nameInput, unitTypeInput, costInput }
  }

  it('renders', () => {
    const { container, unmount } = setup()
    expect(container.innerHTML).toMatchSnapshot()
    unmount()
  })

  it('does not fire submit when incomplete', async () => {
    const { container, unmount, queryByText, onSubmitMock } = setup()

    const submitButton = queryByText('Create ingredient')!

    const buttonClicked = waitForEvent(submitButton, 'click')
    fireEvent.click(submitButton)
    await buttonClicked

    expect(onSubmitMock).not.toHaveBeenCalled()

    expect(container.innerHTML).toMatchSnapshot()

    unmount()
  })

  it('show validation errors', async () => {
    const { container, unmount, queryByText, queryByLabelText, onSubmitMock, submitButton, nameInput, unitTypeInput, costInput } = setup()

    fireEvent.change(nameInput, {target: {value: 'A name'}})
    fireEvent.change(unitTypeInput, {target: {value: 'units'}})
    fireEvent.change(costInput, {target: {value: '-2'}})

    const buttonClicked = waitForEvent(submitButton, 'click')
    fireEvent.click(submitButton)
    await buttonClicked

    expect(onSubmitMock).not.toHaveBeenCalled()

    const errorElement = container.querySelector('.font-bold.text-red-600')
    expect(errorElement).toBeTruthy()
    expect(errorElement!.textContent).toBeTruthy()

    expect(container.innerHTML).toMatchSnapshot()

    unmount()
  })

  it('emits an onSubmit event when valid and submitted', async () => {
    const { unmount, queryByText, queryByLabelText, onSubmitMock, submitButton, nameInput, unitTypeInput, costInput } = setup()

    fireEvent.change(nameInput, {target: {value: 'A name'}})
    fireEvent.change(unitTypeInput, {target: {value: 'units'}})
    fireEvent.change(costInput, {target: {value: '2'}})

    const buttonClicked = waitForEvent(submitButton, 'click')
    fireEvent.click(submitButton)
    await buttonClicked

    expect(onSubmitMock).toHaveBeenCalled()

    unmount()
  })
})
