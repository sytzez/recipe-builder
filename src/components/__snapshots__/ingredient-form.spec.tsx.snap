// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<IngredientForm /> > does not fire submit when incomplete 1`] = `"<form class=\\"mb-4 rounded border border-gray-100 bg-white px-8 py-6 shadow-md\\"><h1 class=\\"mb-4 text-2xl font-bold text-gray-800\\">New ingredient</h1><div class=\\"mb-4\\"><label class=\\"mb-2 block font-bold text-gray-800\\" for=\\"name\\">Name</label><input class=\\"w-full appearance-none rounded border py-2 px-3 leading-tight text-gray-800 shadow focus:border-black focus:outline-none\\" id=\\"name\\" type=\\"text\\" placeholder=\\"Honey\\" required=\\"\\"></div><div class=\\"mb-4\\"><label class=\\"mb-2 block font-bold text-gray-800\\" for=\\"unitType\\">Unit of measurement</label><select class=\\"w-full rounded border py-2 px-3 leading-tight text-gray-800 shadow focus:border-black focus:outline-none\\" id=\\"unitType\\"><option value=\\"units\\">units</option><option value=\\"grams\\">grams</option><option value=\\"ml\\">ml</option><option value=\\"spoons\\">spoons</option></select></div><div class=\\"mb-4\\"><label class=\\"mb-2 block font-bold text-gray-800\\" for=\\"cost\\">Cost per unit of measurement</label><input class=\\"w-full appearance-none rounded border py-2 px-3 leading-tight text-gray-800 shadow focus:border-black focus:outline-none\\" id=\\"cost\\" type=\\"number\\" placeholder=\\"0.05\\" required=\\"\\" step=\\".01\\"></div><div class=\\"flex flex-wrap gap-2\\"><input type=\\"submit\\" class=\\"focus:shadow-outline cursor-pointer rounded-full bg-lime-400 py-2 px-4 font-bold text-white hover:bg-lime-500 focus:outline-none\\" value=\\"Create ingredient\\"><button class=\\"focus:shadow-outline rounded-full bg-gray-100 py-2 px-4 font-bold text-black hover:bg-gray-300 focus:outline-none\\">Cancel</button></div></form>"`;

exports[`<IngredientForm /> > renders 1`] = `"<form class=\\"mb-4 rounded border border-gray-100 bg-white px-8 py-6 shadow-md\\"><h1 class=\\"mb-4 text-2xl font-bold text-gray-800\\">New ingredient</h1><div class=\\"mb-4\\"><label class=\\"mb-2 block font-bold text-gray-800\\" for=\\"name\\">Name</label><input class=\\"w-full appearance-none rounded border py-2 px-3 leading-tight text-gray-800 shadow focus:border-black focus:outline-none\\" id=\\"name\\" type=\\"text\\" placeholder=\\"Honey\\" required=\\"\\"></div><div class=\\"mb-4\\"><label class=\\"mb-2 block font-bold text-gray-800\\" for=\\"unitType\\">Unit of measurement</label><select class=\\"w-full rounded border py-2 px-3 leading-tight text-gray-800 shadow focus:border-black focus:outline-none\\" id=\\"unitType\\"><option value=\\"units\\">units</option><option value=\\"grams\\">grams</option><option value=\\"ml\\">ml</option><option value=\\"spoons\\">spoons</option></select></div><div class=\\"mb-4\\"><label class=\\"mb-2 block font-bold text-gray-800\\" for=\\"cost\\">Cost per unit of measurement</label><input class=\\"w-full appearance-none rounded border py-2 px-3 leading-tight text-gray-800 shadow focus:border-black focus:outline-none\\" id=\\"cost\\" type=\\"number\\" placeholder=\\"0.05\\" required=\\"\\" step=\\".01\\"></div><div class=\\"flex flex-wrap gap-2\\"><input type=\\"submit\\" class=\\"focus:shadow-outline cursor-pointer rounded-full bg-lime-400 py-2 px-4 font-bold text-white hover:bg-lime-500 focus:outline-none\\" value=\\"Create ingredient\\"><button class=\\"focus:shadow-outline rounded-full bg-gray-100 py-2 px-4 font-bold text-black hover:bg-gray-300 focus:outline-none\\">Cancel</button></div></form>"`;

exports[`<IngredientForm /> > show validation errors 1`] = `"<form class=\\"mb-4 rounded border border-gray-100 bg-white px-8 py-6 shadow-md\\"><h1 class=\\"mb-4 text-2xl font-bold text-gray-800\\">New ingredient</h1><div class=\\"mb-4\\"><label class=\\"mb-2 block font-bold text-gray-800\\" for=\\"name\\">Name</label><input class=\\"w-full appearance-none rounded border py-2 px-3 leading-tight text-gray-800 shadow focus:border-black focus:outline-none\\" id=\\"name\\" type=\\"text\\" placeholder=\\"Honey\\" required=\\"\\"></div><div class=\\"mb-4\\"><label class=\\"mb-2 block font-bold text-gray-800\\" for=\\"unitType\\">Unit of measurement</label><select class=\\"w-full rounded border py-2 px-3 leading-tight text-gray-800 shadow focus:border-black focus:outline-none\\" id=\\"unitType\\"><option value=\\"units\\">units</option><option value=\\"grams\\">grams</option><option value=\\"ml\\">ml</option><option value=\\"spoons\\">spoons</option></select></div><div class=\\"mb-4\\"><label class=\\"mb-2 block font-bold text-gray-800\\" for=\\"cost\\">Cost per unit of measurement</label><input class=\\"w-full appearance-none rounded border py-2 px-3 leading-tight text-gray-800 shadow focus:border-black focus:outline-none\\" id=\\"cost\\" type=\\"number\\" placeholder=\\"0.05\\" required=\\"\\" step=\\".01\\"></div><div class=\\"flex flex-wrap gap-2\\"><input type=\\"submit\\" class=\\"focus:shadow-outline cursor-pointer rounded-full bg-lime-400 py-2 px-4 font-bold text-white hover:bg-lime-500 focus:outline-none\\" value=\\"Create ingredient\\"><button class=\\"focus:shadow-outline rounded-full bg-gray-100 py-2 px-4 font-bold text-black hover:bg-gray-300 focus:outline-none\\">Cancel</button></div><p class=\\"font-bold text-red-600\\">cost per unit of measurement must be greater than or equal to 0</p></form>"`;
